doc/routing.md

```diff
@@ -0,0 +1,71 @@
+
+
+Flow Control
+------------
+
+Cockpit's protocol passes messages on reliable, ordered underlying
+transports (eg: WebSockets, HTTP responses, stdio) between multiple
+peers, aranged in in a non-rooted tree graph.
+
+Cockpit's flow control is not about replying messages, but avoiding
+flooding any peer by sending messages along any link in the graph
+too rapidly.
+
+Cockpit's flow control operates at the level of channels in the protocol
+(described elsewhere). A simple blocking of any peer's message input
+would block all channels, and not have the desired effect of slowing
+down a certain channel which is flooding communication.
+
+Therefore we use a windowing flow control protocol. It is based around
+per-channel "ping" control messages. In the Cockpit protocol "ping"
+messages in a channel are responded to by specific leafs of the graph.
+They are replied to with a "pong" contaning otherwise identical content.
+Other "ping" messages are used as keep alives, but these are beyond the
+scope of this description.
+
+Each channel leaf that participates in flow control sends "ping"
+messages with a sequence number, and waits to see the corresponding
+"pong" message with the same sequence and channel number. By checking
+the "ping" sequence numbers responded to, a channel can slow input and/or
+transmission to wait for opposite leaf to catch up.
+
+Sadly this is not enough. The leaf's in of Cockpit's tree often do not
+consume data themselves, but pass it on to other sockets, streams, files
+and so on.
+
+Various leaf's monitor their output edge queues and hold back responding
+to "ping" messages if their output edge queue is too full. This is done
+by use of "pressure" signals on the output queue to indicate when pressure
+is high or low.
+
+This results in flow control that is achieves its intent, but does not
+make guarantees. In addition only some channels and leafs implement flow
+control at this point:
+
+ * cockpit.js: Responds to "ping" messages indicating that a given
+   sequence has reached the browser.
+
+ * CockpitChannel: Sends sequence pings, every so often in the channel.
+   Slows down an input pipe via a "pressure" signal when too many
+   sequences are outstanding and have not been responded to.
```

Comment:
CockpitChannel also appears to be responsible for responding to the pings, and holding back responses when it has been throttled by a pressure signal from another object.